File: DivideNConquere/tiling.py
import sys
kb = sys.stdin


def tiling(a,b,L,missing):
    #basecase
    X,Y = missing
    #print("X-Y",X,Y)
    # Base case: If the current tile is 2x2
    #if (x2 - x1 == 1) and (y2 - y1 == 1):
    if (L == 2):
        dx=X-a
        dy=Y-b
        p=2*dy+dx
        print(p, a, b)
        return
    #check which Q of missing?
    mid=L//2
    
    posX=a+mid-1
    posY=b+mid-1
    if(X<a+mid and Y<b+mid):#Q1
        print(0,posX,posY)
        #tiling(a,b,L,missing)
        tiling(a,b,mid,missing) #Q1
        tiling(a+mid,b,mid,(a+mid,b+mid-1))#Q2
        tiling(a,b+mid,mid,(a+mid-1,b+mid))#Q3
        tiling(a+mid,b+mid,mid,(a+mid,b+mid))#Q4
    elif(X>=a+mid and Y<b+mid):#Q2
        print(1,posX,posY)
        tiling(a,b,mid,(a+mid-1,b+mid-1)) #Q1
        tiling(a+mid,b,mid,missing)#Q2
        tiling(a,b+mid,mid,(a+mid-1,b+mid))#Q3
        tiling(a+mid,b+mid,mid,(a+mid,b+mid))#Q4
    elif(X<a+mid and Y>=b+mid):#Q3
        print(2,posX,posY)
        tiling(a,b,mid,(a+mid-1,b+mid-1)) #Q1
        tiling(a+mid,b,mid,(a+mid,b+mid-1))#Q2
        tiling(a,b+mid,mid,(missing))#Q3
        tiling(a+mid,b+mid,mid,(a+mid,b+mid))#Q4
    else:#Q4
        print(3,posX,posY)
        tiling(a,b,mid,(a+mid-1,b+mid-1)) #Q1
        tiling(a+mid,b,mid,(a+mid,b+mid-1))#Q2
        tiling(a,b+mid,mid,(a+mid-1,b+mid))#Q3
        tiling(a+mid,b+mid,mid,(missing))#Q4
    
    #recur
L,x,y=[ int(e) for e in kb.readline().split()]
print(L*L//3)
tiling(0,0,L,(x,y))
""" tiling(m,n,L,(2,0))
tiling(m,n,L,(3,0))
tiling(m,n,L,(2,1))
tiling(m,n,L,(3,1)) """

File: DivideNConquere/tiling.cpp
#include <iostream>
#include <vector>
using namespace std;

// A utility structure to represent a tile or a missing position
struct Tile {
    int type, x, y;
    Tile(int type, int x, int y) : type(type), x(x), y(y) {}
};

vector<Tile> answer;

void tiling(int x1, int x2, int y1, int y2, pair<int, int> missing) {
    // Base case: If the current tile is 1x1
    if ((x2 - x1 == 1) && (y2 - y1 == 1)) {
        int tile_mapping[2][2] = {{0, 1}, {2, 3}};
        answer.push_back(Tile(tile_mapping[missing.first - x1][missing.second - y1], x1, y1));
        return;
    }
    // Determine the midpoint
    int mid_x = (x1 + x2) / 2;
    int mid_y = (y1 + y2) / 2;
    int X = missing.first;
    int Y = missing.second;
    // Determine which quadrant the missing tile is in and recurse appropriately
    if (X < mid_x && Y < mid_y) { // Q1
        answer.push_back(Tile(0, mid_x, mid_y));
        tiling(x1, mid_x, y1, mid_y, missing);
    } else if (X >= mid_x && Y < mid_y) { // Q2
        answer.push_back(Tile(1, mid_x, mid_y));
        tiling(mid_x, x2, y1, mid_y, missing);
    } else if (X < mid_x && Y >= mid_y) { // Q3
        answer.push_back(Tile(2, mid_x, mid_y));
        tiling(x1, mid_x, mid_y, y2, missing);
    } else { // Q4
        answer.push_back(Tile(3, mid_x, mid_y));
        tiling(mid_x, x2, mid_y, y2, missing);
    }

    // Fill the other quadrants with a fake missing tile at their center
    if (missing != make_pair(mid_x, mid_y)) tiling(x1, mid_x, y1, mid_y, {mid_x - 1, mid_y - 1});
    if (missing != make_pair(mid_x + 1, mid_y)) tiling(mid_x, x2, y1, mid_y, {mid_x, mid_y - 1});
    if (missing != make_pair(mid_x, mid_y + 1)) tiling(x1, mid_x, mid_y, y2, {mid_x - 1, mid_y});
    if (missing != make_pair(mid_x + 1, mid_y + 1)) tiling(mid_x, x2, mid_y, y2, {mid_x, mid_y});
}

int main() {
    int n, x, y;
    cin >> n >> x >> y;

    // Recursively tile the given square
    tiling(0, n, 0, n, {x, y});

    cout << answer.size() << endl;
    for (const auto &tile : answer) {
        cout << tile.type << " " << tile.x << " " << tile.y << endl;
    }

    return 0;
}
